src/box/memtx_tx.c

```diff
@@ -2032,6 +2142,18 @@ memtx_tx_history_add_insert_stmt(struct txn_stmt *stmt,
 	for (uint32_t i = 0; i < space->index_count; i++) {
 		struct tuple *next = directly_replaced[i];
 		struct tuple *succ = direct_successor[i];
+		/*
+		 * The tuple was inserted as new according to the operating
+		 * transaction and it's not excluded from the index. It's
+		 * possible that the operation replaced a tuple in the index,
+		 * but, if so, the tuple is not visible to the inserter.
+		 */
+		if (old_tuple == NULL && !tuple_excluded[i])
+			memtx_tx_handle_counted_write(space, add_story, i);
```

Comment:
If write happened before count, it's not handled - you need to handle counted writes when they are prepared as well.
I already have a test case for you:
```lua
g_mvcc.test_review = function()
    g_mvcc.server:exec(function()
        -- The test space.
        local s = box.schema.space.create('test')
        s:create_index('pk', {fast_offset = true,
                              parts = {{1, 'unsigned'}, {2, 'unsigned'}}})
            
        -- Prepare proxies.
        local txn_proxy = require('test.box.lua.txn_proxy')
        local tx1 = txn_proxy.new()
        local tx2 = txn_proxy.new()

        tx1:begin()
        tx2:begin()
        tx1('box.space.test:replace{1, 0}')
        
        -- No prepared tuples - count must return 0
        local count = tx2('return box.space.test:count{1, 0}')[1]
        t.assert_equals(count, 0)

        -- Commit writer
        tx1:commit()

        -- Count again - the same value (zero) must be returned
        count = tx2('return box.space.test:count{1, 0}')[1]
        t.assert_equals(count, 0)
        tx2:commit()
    
        -- Check if insert actually happened
        t.assert_equals(box.space.test:select{}, {{1, 0}})
    end)
end
```

Now it fails on the second check of `count`.

Reply:
Fixed.